### PR TITLE
Add friendly error message to UrlHelper for when there is no IRouter

### PR DESCRIPTION
**Pull request by JamesNK**
Addresses #7283

### CODE REVIEW

File: src/Mvc/Mvc.Core/src/Routing/UrlHelper.cs [hunk 1]
```diff
@@ -32,7 +32,21 @@ public UrlHelper(ActionContext actionContext)
...
         /// </summary>
-        protected IRouter Router => ActionContext.RouteData.Routers[0];
+        protected IRouter Router
+        {
+            get
+            {
+                var routers = ActionContext.RouteData.Routers;
+                if (routers.Count == 0)
+                {
+                    throw new InvalidOperationException("Could not find an IRouter associated with the ActionContext. "
```
Comment (JamesNK) [inline src/Mvc/Mvc.Core/src/Routing/UrlHelper.cs]: I also tried this in the UrlHelper constructor and it broke tests. Placing the check here means that the exception will only be thrown if the IndexOutOfRangeException was already going to be thrown.

Comment (rynowak) [reply src/Mvc/Mvc.Core/src/Routing/UrlHelper.cs]: Sounds good.

File: src/Mvc/Mvc.Core/src/Routing/UrlHelper.cs [hunk 2]
```diff
@@ -32,7 +32,21 @@ public UrlHelper(ActionContext actionContext)
...
+                        + "If your application is using endpoint routing then you can get a UrlHelper instance with "
+                        + "dependency injection or use Microsoft.AspNetCore.Routing.LinkGenerator.");
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Routing/UrlHelper.cs]: This is false. You can't get a `UrlHelper` from DI. That's one of the problems that link generator solves.

users can instead resolve an `IUrlHelperFactory` and put in the action context. Don't have an action context? that's another one of the problems link generator solves.

**Comment by JamesNK** [comment]
:up: :date:

**Review by rynowak — APPROVED** [review]

**Pull request merged**